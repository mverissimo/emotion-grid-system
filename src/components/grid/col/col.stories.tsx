import type { Meta, StoryObj } from '@storybook/react-vite';

import Col from './col';

const meta = {
  component: Col,
  title: 'Components/Grid/Col',
} satisfies Meta<typeof Col>;

export const Base: StoryObj<typeof meta> = {
  args: {
    size: {
      md: 1,
    },
  },
};

export default meta;
