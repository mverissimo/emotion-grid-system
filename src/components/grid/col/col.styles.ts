import { css, cx } from '@emotion/css';
import type { Theme } from '@emotion/react';

import { mediaOnly, responsive } from '../../../utils/media';
import type { ColProps } from './col';

const initial = (theme: Theme) =>
  cx(
    css({
      label: 'grid-col',

      flex: '1 0 0%',
      boxSizing: 'border-box',

      width: '100%',
      maxWidth: '100%',
    }),
    responsive(theme, (breakpoint) => ({
      paddingLeft: `calc(${theme.grid?.gutter[breakpoint]} / 2)`,
      paddingRight: `calc(${theme.grid?.gutter[breakpoint]} / 2)`,
    })),
  );

const size = (theme: Theme, size: ColProps['size']) => {
  if (size) {
    return cx(
      css({
        label: 'grid-col--size',
      }),
      mediaOnly(theme, size, (breakpoint) => ({
        flex: `${size[breakpoint] === 'auto' ? 'auto' : `0 0 calc((${size[breakpoint]} / ${theme.grid?.columns[breakpoint]}) * 100%)`}`,
      })),
    );
  }
};

const offset = (theme: Theme, offset: ColProps['offset']) => {
  if (offset) {
    return cx(
      css({
        label: 'grid-col--offset',
      }),
      mediaOnly(theme, offset, (breakpoint) => ({
        marginLeft: `calc((${offset[breakpoint]} / ${theme.grid?.columns[breakpoint]}) * 100%)`,
      })),
    );
  }
};

export default {
  initial,
  size,
  offset,
};
