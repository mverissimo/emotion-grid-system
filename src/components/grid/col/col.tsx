import type { HTMLAttributes, Ref } from 'react';

import { cx } from '@emotion/css';
import type { Breakpoint } from '../../../theme';

import { useTheme } from '../../../hooks/use-theme';

import styles from './col.styles';

export interface ColProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Set the column size
   */
  size?: Partial<Record<Breakpoint, number | 'auto'>>;

  /**
   * Set the column offset
   */
  offset?: Partial<Record<Breakpoint, number>>;
}

function Col(props: ColProps, ref?: Ref<HTMLDivElement>) {
  const { size, offset, className, ...rest } = props;

  const theme = useTheme();

  return (
    <div
      data-grid="col"
      className={cx(
        styles.initial(theme),
        styles.size(theme, size),
        styles.offset(theme, offset),
        className,
      )}
      ref={ref}
      {...rest}
    />
  );
}

export default Col;
