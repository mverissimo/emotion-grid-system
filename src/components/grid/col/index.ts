export { default as Col } from './col';
export type { ColProps } from './col';
