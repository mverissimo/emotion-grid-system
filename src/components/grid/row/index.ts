export { default as Row } from './row';
export type { RowProps } from './row';
