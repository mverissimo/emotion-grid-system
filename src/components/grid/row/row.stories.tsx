import type { Meta, StoryObj } from '@storybook/react-vite';

import Row from './row';

const meta = {
  component: Row,
  title: 'Components/Grid/Row',
} satisfies Meta<typeof Row>;

export const Base: StoryObj<typeof meta> = {};

export default meta;
