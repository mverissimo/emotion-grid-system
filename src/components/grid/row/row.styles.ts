import { css, cx } from '@emotion/css';
import type { Theme } from '@emotion/react';

import { mediaOnly, responsive } from '../../../utils/media';
import type { RowProps } from './row';

const base = (theme: Theme) =>
  cx(
    css({
      label: 'grid-row',

      display: 'flex',
      flexWrap: 'wrap',

      boxSizing: 'border-box',
    }),
    responsive(theme, (breakpoint) => ({
      marginLeft: `calc((${theme.grid?.gutter?.[breakpoint]} / 2) * -1)`,
      marginRight: `calc((${theme.grid?.gutter[breakpoint]} / 2) * -1)`,
    })),
  );

const align = (theme: Theme, align: RowProps['align']) => {
  if (align) {
    return cx(
      css({
        label: 'grid-row--align',
      }),
      mediaOnly(theme, align, (breakpoint) => ({
        alignItems: align[breakpoint],
      })),
    );
  }
};

const justify = (theme: Theme, justify: RowProps['justify']) => {
  if (justify) {
    return cx(
      css({
        label: 'grid-row--justify',
      }),
      mediaOnly(theme, justify, (breakpoint) => ({
        justifyContent: justify[breakpoint],
      })),
    );
  }
};

const reverse = (theme: Theme, reverse: RowProps['reverse']) => {
  if (reverse) {
    return cx(
      css({
        label: 'grid-row--reverse',
      }),
      mediaOnly(theme, reverse, (breakpoint) => ({
        flexDirection: reverse[breakpoint] ? 'row-reverse' : 'row',
      })),
    );
  }
};

const noGutters = (noGutters: RowProps['noGutters']) => {
  if (noGutters) {
    return css({
      '[data-grid=col]': {
        paddingLeft: '0 !important',
        paddingRight: '0 !important',
      },
    });
  }
};

export default {
  base,
  align,
  justify,
  reverse,
  noGutters,
};
