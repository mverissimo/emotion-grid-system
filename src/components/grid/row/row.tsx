import type { CSSProperties, HTMLAttributes, Ref } from 'react';

import { cx } from '@emotion/css';
import type { ResponsiveValue } from '../../../theme';

import { useTheme } from '../../../hooks/use-theme';

import styles from './row.styles';

export interface RowProps extends HTMLAttributes<HTMLDivElement> {
  /**
   *
   */
  align?: ResponsiveValue<CSSProperties['alignItems']>;

  /**
   *
   */
  justify?: ResponsiveValue<CSSProperties['justifyContent']>;

  /**
   *
   */
  reverse?: ResponsiveValue<boolean>;

  /**
   *
   */
  noGutters?: boolean;
}

function Row(props: RowProps, ref?: Ref<HTMLDivElement>) {
  const { align, justify, reverse, noGutters, className, ...rest } = props;

  const theme = useTheme();

  return (
    <div
      data-grid="row"
      className={cx(
        styles.base(theme),
        styles.align(theme, align),
        styles.justify(theme, justify),
        styles.reverse(theme, reverse),
        styles.noGutters(noGutters),
        className,
      )}
      ref={ref}
      {...rest}
    />
  );
}

export default Row;
