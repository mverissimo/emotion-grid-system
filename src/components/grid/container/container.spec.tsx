import { render } from '@testing-library/react';

import { Container } from '.';
import { Col } from '../col';
import { Row } from '../row';

describe('@Grid/Container', () => {
  it('should render with default styles', () => {
    const { container } = render(<Container>Container</Container>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with fluid styles when passed the isFluid prop', () => {
    const { container } = render(
      <Container isFluid={true}>Container fluid</Container>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with debug styles when passed the isDebug prop', () => {
    const { container } = render(
      <Container isDebug={true}>
        <Row>
          <Col>Col</Col>
        </Row>
      </Container>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with a custom className', () => {
    const { container } = render(
      <Container className="my-custom-class-name">Container</Container>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
