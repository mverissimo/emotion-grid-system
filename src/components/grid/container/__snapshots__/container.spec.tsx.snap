// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@Grid/Container > should render with a custom className 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 100%;
  max-width: 100%;
  margin-left: auto;
  margin-right: auto;
  max-width: 100%;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

@media (min-width: 40rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    max-width: 48rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    max-width: 64rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    max-width: 80rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    max-width: 96rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

<div
  class="emotion-0 my-custom-class-name"
  data-grid="container"
>
  Container
</div>
`;

exports[`@Grid/Container > should render with debug styles when passed the isDebug prop 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 100%;
  max-width: 100%;
  margin-left: auto;
  margin-right: auto;
  max-width: 100%;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

@media (min-width: 40rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    max-width: 48rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    max-width: 64rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    max-width: 80rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    max-width: 96rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

.emotion-0 [data-grid=row] {
  background: #2563eb0D;
}

.emotion-0 [data-grid=col] {
  background: #2563eb0D;
  border-width: 1px;
  border-style: solid;
  border-color: #2563eb;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: calc((2rem / 2) * -1);
  margin-right: calc((2rem / 2) * -1);
}

@media (min-width: 40rem) {
  .emotion-1 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 48rem) {
  .emotion-1 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 64rem) {
  .emotion-1 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 80rem) {
  .emotion-1 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

@media (min-width: 96rem) {
  .emotion-1 {
    margin-left: calc((2rem / 2) * -1);
    margin-right: calc((2rem / 2) * -1);
  }
}

.emotion-2 {
  -webkit-flex: 1 0 0%;
  -ms-flex: 1 0 0%;
  flex: 1 0 0%;
  box-sizing: border-box;
  width: 100%;
  max-width: 100%;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

@media (min-width: 40rem) {
  .emotion-2 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 48rem) {
  .emotion-2 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 64rem) {
  .emotion-2 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 80rem) {
  .emotion-2 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 96rem) {
  .emotion-2 {
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

<div
  class="emotion-0"
  data-grid="container"
>
  <div
    class="emotion-1"
    data-grid="row"
  >
    <div
      class="emotion-2"
      data-grid="col"
    >
      Col
    </div>
  </div>
</div>
`;

exports[`@Grid/Container > should render with default styles 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 100%;
  max-width: 100%;
  margin-left: auto;
  margin-right: auto;
  max-width: 100%;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

@media (min-width: 40rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    max-width: 48rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    max-width: 64rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    max-width: 80rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    max-width: 96rem;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

<div
  class="emotion-0"
  data-grid="container"
>
  Container
</div>
`;

exports[`@Grid/Container > should render with fluid styles when passed the isFluid prop 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 100%;
  max-width: 100%;
  margin-left: auto;
  margin-right: auto;
  max-width: 100%;
  padding-left: calc(2rem / 2);
  padding-right: calc(2rem / 2);
}

@media (min-width: 40rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 48rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 64rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

@media (min-width: 96rem) {
  .emotion-0 {
    max-width: 100%;
    padding-left: calc(2rem / 2);
    padding-right: calc(2rem / 2);
  }
}

<div
  class="emotion-0"
  data-grid="container"
>
  Container fluid
</div>
`;
